allennlp/data/semparse/worlds/nlvr_world.py

```diff
@@ -0,0 +1,509 @@
+
+            2a) Object Attribute Functions: They take sets of objects and return sets of
+            attributes.  `color`, `shape` are the attribute functions.
+
+            2b) Count Function : Takes a set of objects or boxes and returns its cardinality.
+
+        3) Box Membership Function : This takes a box as an argument and returns the objects in it.
+        This is a special kind of attribute function for two reasons. Firstly, it returns a set of
+        objects instead of attributes, and secondly it occurs only within the second argument of a
+        box filtering function (see below). It provides a way to query boxes based on the
+        attributes of objects contained within it. The function is called `object_in_box`, and it
+        gets executed within ``_execute_box_filter``.
+
+        4) Box Filtering Functions : These are of the form `filter(set_of_boxes,
+        attribute_function, target_attribute)` The idea is that we take a set of boxes, an
+        attribute function that extracts the relevant attribute from a box, and a target attribute
+        that we compare against. The logic is that we execute the attribute function on _each_ of
+        the given boxes and return only those whose attribute value, in comparison with the target
+        attribute, satisfies the filtering criterion (i.e., equal to the target, less than, greater
+        than etc.). The fitering function defines the comparison operator.  All the functions in
+        this class with names `filter_*` belong to this category.
+
+        5) Object Filtering Functions : These are of the form `filter(set_of_objects)`. These are
+        similar to box filtering functions, but they operate on objects instead. Also, note that
+        they take just one argument instead of three. This is because while box filtering functions
+        typically query complex attributes, object filtering functions query the properties of the
+        objects alone.  These are simple and finite in number. Thus, we essentially let the
+        filtering function define the attribute function, and the target attribute as well, along
+        with the comparison operator.  That is, these are functions like `black` (which takes a set
+        of objects, and returns those whose "color" (attribute function) "equals" (comparison
+        operator) "black" (target attribute)), or "square" (which returns objects that are
+        squares).
+
+        6) Negate Object Filter : Takes an object filter and a set of objects and applies the
+        negation of the object filter on the set.
+        """
+        if not logical_form.startswith("("):
+            logical_form = "(%s)" % logical_form
+        logical_form = logical_form.replace(",", " ")
+        expression_as_list = pyparsing.OneOrMore(pyparsing.nestedExpr()).parseString(logical_form).asList()
+        # The whole expression has to be an assertion expression because it has to return a boolean.
```

Comment:
Good point. It wouldn't be too hard to generalize, but I may have to change the structure of the high level method to be able to deal with questions and assertions differently, or may be have two different high level methods. That said, I wouldn't say this is too specific either, because we're only assuming the input is an assertion.
I added a TODO for this.